tiledb/sm/query/readers/attribute_order_validator.h

```diff
@@ -0,0 +1,581 @@
+      auto max = non_empty_domain[1];
+      auto max_tile_idx = fragment_metadata[f]->tile_num() - 1;
+      bool max_tile_aligned = max + 1 == index_dim->round_to_tile<IndexType>(
+                                             max + 1, dim_dom[0], tile_extent);
+
+      // Get the tile index of the tile in f2 immediately following f (with
+      // respect to tile index in f2). If f is tile aligned we need to add an
+      // additional value.
+      uint64_t f2_tile_idx = max_tile_idx + frag_first_array_tile_idx[f] -
+                             frag_first_array_tile_idx[f2] + max_tile_aligned;
+
+      // If we are tile aligned or the max is right next to the other fragment's
+      // min, we can validate. Otherwise we'll need to load the tile.
+      const IndexType* non_empty_domain2 =
+          static_cast<const IndexType*>(non_empty_domains[f2]);
+      if (max_tile_aligned || max + 1 == non_empty_domain2[0]) {
+        max_validated_[f] = true;
+
+        // Check the order.
+        if (increasing_data) {
+          // Increasing data: The last value on f is the maximum on f. Check
+          // that is less than the first (minimum) value on the following
+          // tile in f2.
+          auto value = fragment_metadata[f]->get_tile_max_as<AttributeType>(
+              attribute_name_, max_tile_idx);
+          auto value_next =
+              fragment_metadata[f2]->get_tile_min_as<AttributeType>(
+                  attribute_name_, f2_tile_idx);
+          if (value_next <= value) {
+            throw AttributeOrderValidatorStatusException(
+                "Attribute out of order");
+          }
+
+        } else {
+          // Decreasinging data: The last value on f is the minimum on f. Check
+          // that is greater than the first (maximum) value on the following
+          // tile in f2.
+          auto value = fragment_metadata[f]->get_tile_min_as<AttributeType>(
+              attribute_name_, max_tile_idx);
+          auto value_next =
+              fragment_metadata[f2]->get_tile_max_as<AttributeType>(
+                  attribute_name_, f2_tile_idx);
+          if (value_next >= value) {
+            throw AttributeOrderValidatorStatusException(
+                "Attribute out of order");
+          }
+        }
+
+      } else {
+        // Add the tile to the list of tiles to load.
+        add_tile_to_load(f, false, f2, f2_tile_idx, schema);
+      }
+    }
+  }
+
+  /**
+   * Performs validation that requires tile data.
+   *
+   * For best performance, only execute this method after first validating
+   * without loading tiles. If a validation check fails, this will throw an
+   * error.
+   *
+   * @tparam Index type.
+   * @tparam Attribute type.
+   * @param index_dim The index dimension.
+   * @param f Fragment index.
+   * @param non_empty_domains Vector of pointers to the non-empty domain for
+   *     each fragment.
+   *
+   */
```

Comment:
```suggestion
  /**
   * Performs validation that requires tile data.
   *
   * For best performance, only execute this method after first validating
   * without loading tiles.  This requires ``find_fragments_to_check`` to be
   * ran prior to execution.  If a validation check fails, this will throw an
   * error.
   *
   * @tparam Index type.
   * @tparam Attribute type.
   * @param index_dim The index dimension.
   * @param f Fragment index.
   * @param non_empty_domains Vector of pointers to the non-empty domain for
   *     each fragment.
   */
```